prow/tide/tide.go

```diff
@@ -835,15 +836,38 @@ func (c *Controller) mergePRs(sp subpool, prs []PullRequest) error {
 	return nil
 }
 
+func aggregateChangeProvider(providers []config.ChangedFilesProvider) config.ChangedFilesProvider {
+	return func() ([]string, error) {
+		var changes []string
+		var errs []error
+		for _, provider := range providers {
+			specificChanges, specificErr := provider()
+			changes = append(changes, specificChanges...)
+			errs = append(errs, specificErr)
+		}
+		return changes, errorutil.NewAggregate(errs...)
+	}
+}
+
 func (c *Controller) trigger(sp subpool, presubmitContexts map[int]sets.String, prs []PullRequest) error {
 	requiredContexts := sets.NewString()
 	for _, pr := range prs {
 		requiredContexts = requiredContexts.Union(presubmitContexts[int(pr.Number)])
 	}
 
-	// TODO(cjwagner): DRY this out when generalizing triggering code (and code to determine required and to-run jobs).
 	for _, ps := range c.ca.Config().Presubmits[sp.org+"/"+sp.repo] {
-		if ps.SkipReport || !ps.RunsAgainstBranch(sp.branch) || !requiredContexts.Has(ps.Context) {
+		if !requiredContexts.Has(ps.Context) || !ps.ContextRequired() {
+			continue
+		}
+		var providers []config.ChangedFilesProvider
+		for _, pr := range prs {
+			providers = append(providers, config.NewGitHubDeferredChangedFilesProvider(c.ghc, sp.org, sp.repo, int(pr.Number)))
+		}
+		shouldRun, err := ps.ShouldRun(sp.branch, "", aggregateChangeProvider(providers))
```

Comment:
This isn't the same as `tide` used to not consider jobs that should not trigger based on `run_if_changed`?

Reply:
This is incorrect. This is handled by the `presubmitsByPull` function: if the job needs to run as a result of a `run_if_changed` rule, the context will be included in `requiredContexts`.

Reply:
It seems wrong to do the shouldRun checks twice, but I think we need it to distinguish sharded jobs unless we collect the actual presubmit configs instead of just the context names in `presubmitsByPull` (that might be the best option).